### PR TITLE
[libclc] Move min/max/clamp into the CLC builtins library

### PR DESCRIPTION
**Pull request by frasercrmck**
These functions are "shared" between integer and floating-point types, hence the directory name. They are used in several CLC internal functions such as __clc_ldexp.

Note that clspv and spirv targets don't want to define these functions, so pre-processor macros replace calls to __clc_min with regular min, for example. This means they can use as much of the generic CLC source files as possible, but where CLC functions would usually call out to an external __clc_min symbol, they call out to an external min symbol. Then they opt out of defining __clc_min itself in their CLC builtins library.

Preprocessor definitions for these targets have also been changed somewhat: what used to be CLC_SPIRV (the 32-bit target) is now CLC_SPIRV32, and CLC_SPIRV now represents either CLC_SPIRV32 or CLC_SPIRV64. Same goes for CLC_CLSPV.

There are no differences (measured with llvm-diff) in any of the final builtins libraries for nvptx, amdgpu, or clspv. Neither are there differences in the SPIR-V targets' LLVM IR before it's actually lowered to SPIR-V.

### CODE REVIEW

**Comment by frasercrmck** [comment]
CC @rjodinchr, I hope this method of dealing with clspv works for you.

File: libclc/clc/lib/generic/shared/clc_max.inc [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+                                               __CLC_GENTYPE b) {
+  return (a > b ? a : b);
+}
+
+#ifndef __CLC_SCALAR
+_CLC_OVERLOAD _CLC_DEF __CLC_GENTYPE __clc_max(__CLC_GENTYPE a,
+                                               __CLC_SCALAR_GENTYPE b) {
+  return (a > (__CLC_GENTYPE)b ? a : (__CLC_GENTYPE)b);
+}
+#endif
```
Comment (arsenm) [inline libclc/clc/lib/generic/shared/clc_max.inc:11]: Why not just use __builtin_elementwise_max? It's simplest to just not have all this boilerplate

Comment (frasercrmck) [reply libclc/clc/lib/generic/shared/clc_max.inc:11]: Good idea, seems like a sensible refactor while we're at it. I'm generally just copy/pasting code over to try and affect as little as possible.

**Review by arsenm — CHANGES_REQUESTED** [review]
Moving them seems fine but these should probably just be deleted. __clc_min is equivalent to __builtin_elementwise_min etc.

I see there's a __builtin_hlsl_elementwise_clamp, but given the language prefix it may be more abusive

**Review by arsenm — APPROVED** [review]

**Comment by frasercrmck** [comment]
> Moving them seems fine but these should probably just be deleted. __clc_min is equivalent to __builtin_elementwise_min etc.
> 
> I see there's a __builtin_hlsl_elementwise_clamp, but given the language prefix it may be more abusive

**Comment by arsenm** [comment]
We could probably extend the elementwise builtins to handle the splat case

**Comment by frasercrmck** [comment]
> We could probably extend the elementwise builtins to handle the splat case

Yeah I don't see why not. Let me see what I can do. If you're okay with it, I might just merge this as it is and come back to it if and when the builtins are improved.

**Review by arsenm — APPROVED** [review]

**Pull request merged**